zrml/prediction-markets/src/lib.rs

```diff
@@ -627,53 +617,59 @@ mod pallet {
-                &Self::reserve_id(),
-                Asset::Ztg,
-                &who,
-                default_dispute_bond::<T>(disputes.len()),
-            )?;
-            // TODO(#782): use multiple benchmarks paths for different dispute mechanisms
-            match market.dispute_mechanism {
+            ensure!(market.status == MarketStatus::Reported, Error::<T>::InvalidMarketStatus);
+
+            let overweight = match market.dispute_mechanism {
                 MarketDisputeMechanism::Authorized => {
-                    T::Authorized::on_dispute(&disputes, &market_id, &market)?
+                    T::Authorized::on_dispute(&market_id, &market)?;
+                    T::Court::on_dispute_max_weight()
+                        .saturating_add(T::SimpleDisputes::on_dispute_max_weight())
                 }
                 MarketDisputeMechanism::Court => {
-                    T::Court::on_dispute(&disputes, &market_id, &market)?
+                    T::Court::on_dispute(&market_id, &market)?;
+                    T::Authorized::on_dispute_max_weight()
+                        .saturating_add(T::SimpleDisputes::on_dispute_max_weight())
                 }
                 MarketDisputeMechanism::SimpleDisputes => {
-                    T::SimpleDisputes::on_dispute(&disputes, &market_id, &market)?
+                    T::SimpleDisputes::on_dispute(&market_id, &market)?;
+                    T::Court::on_dispute_max_weight()
+                        .saturating_add(T::Authorized::on_dispute_max_weight())
                 }
-            }
+            };
+
+            let dispute_bond = T::DisputeBond::get();
+
+            T::AssetManager::reserve_named(&Self::reserve_id(), Asset::Ztg, &who, dispute_bond)?;
 
-            Self::set_market_as_disputed(&market, &market_id)?;
-            let market_dispute = MarketDispute { at: curr_block_num, by: who, outcome };
-            <Disputes<T>>::try_mutate(market_id, |disputes| {
-                disputes.try_push(market_dispute.clone()).map_err(|_| <Error<T>>::StorageOverflow)
+            <zrml_market_commons::Pallet<T>>::mutate_market(&market_id, |m| {
+                m.status = MarketStatus::Disputed;
+                m.bonds.dispute = Some(Bond::new(who.clone(), dispute_bond));
+                Ok(())
             })?;
-            Self::deposit_event(Event::MarketDisputed(
-                market_id,
-                MarketStatus::Disputed,
-                market_dispute,
-            ));
-            Ok((Some(T::WeightInfo::dispute_authorized())).into())
+
+            Self::deposit_event(Event::MarketDisputed(market_id, MarketStatus::Disputed));
+
+            let full_weight = T::WeightInfo::dispute_authorized().saturating_add(
+                T::Court::on_dispute_max_weight()
+                    .saturating_add(T::SimpleDisputes::on_dispute_max_weight()),
+            );
+
+            Ok((Some(full_weight.saturating_sub(overweight))).into())
```

Comment:
I think we already reached much better precision in regards to the weights. One last thing I am missing here is that we use the actual weight instead of the maximum weight.

This is the commit: https://github.com/sea212/zeitgeist/commit/baaadda7407de7636b9426027e3c25fa56dbf76d
Instructions how to include it: https://stackoverflow.com/a/13789405/10273801